File: Offer 41. 数据流中的中位数.java
class MedianFinder {
    PriorityQueue<Integer> small, big;

    /** initialize your data structure here. */
    public MedianFinder() {
        small = new PriorityQueue<>((x, y) -> (y - x)); // 小的半边为大顶堆
        big = new PriorityQueue<>(); // 大的半边为小顶堆
    }
    
    public void addNum(int num) {
        if (small.size() == big.size()) {
            // 偶数
            big.offer(num);
            small.offer(big.poll());
        } else {
            // 奇数
            small.offer(num);
            big.offer(small.poll());
        }
    }
    
    public double findMedian() {
        return (small.size() == big.size())? (small.peek() + big.peek()) / 2.0 : small.peek();
    }
}

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder obj = new MedianFinder();
 * obj.addNum(num);
 * double param_2 = obj.findMedian();
 */

File: Offer 41. 数据流中的中位数.py
class MedianFinder:

    def __init__(self):
        """
        initialize your data structure here.
        """

        # 左边是大顶堆，右边是小顶堆
        self.left, self.right = [], []


    def addNum(self, num: int) -> None:
        if len(self.left) == len(self.right):
            heapq.heappush(self.left, -num)
            heapq.heappush(self.right, -heapq.heappop(self.left))
        else:
            heapq.heappush(self.right, num)
            heapq.heappush(self.left, -heapq.heappop(self.right))


    def findMedian(self) -> float:
        return (self.right[0] - self.left[0]) / 2 if len(self.left) == len(self.right) else self.right[0]


class MedianFinder:

    def __init__(self):
        """
        initialize your data structure here.
        """

        # 左边是大顶堆，右边是小顶堆
        self.left, self.right = [], []


    def addNum(self, num: int) -> None:
        if len(self.left) == len(self.right):
            heapq.heappush(self.right, -heapq.heappushpop(self.left, -num))
        else:
            heapq.heappush(self.left, -heapq.heappushpop(self.right, num))


    def findMedian(self) -> float:
        return (self.right[0] - self.left[0]) / 2 if len(self.left) == len(self.right) else self.right[0]


# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()